sympy/solvers/solveset.py

```diff
@@ -2761,15 +2871,34 @@ def linsolve(system, *symbols):
     EmptySet
 
     * An error is raised if, after expansion, any nonlinearity
-      is detected:
+      is detected.
 
-    >>> linsolve([x*(1/x - 1), (y - 1)**2 - y**2 + 1], x, y)
-    {(1, 1)}
-    >>> linsolve([x**2 - 1], x)
+    >>> linsolve([x*(x + 1) - 1], x)
+    Traceback (most recent call last):
+    ...
+    NonlinearError:
+    symbol-dependent cross-terms encountered
+
+    In some cases the message will indicate that using ``strict=False``
+    might be able to ignore the factor:
+
+    >>> xyeqs = [x*(1/x - 1), (y - 1)**2 - y**2 + 1]
+    >>> linsolve(xyeqs, x, y)
     Traceback (most recent call last):
     ...
     NonlinearError:
-    nonlinear term encountered: x**2
+    symbol-dependent term can be ignored using `strict=False`
+
+    But use ``strict=False`` with caution since the results may need
+    nuanced interpretation:
+
+    >>> linsolve(xyeqs, x, y, strict=False)
+    EmptySet
+    >>> e1, e2 = xyeqs
+    >>> linsolve([e1, e2.expand()], x, y, strict=False)
+    {(0, 1)}
+    >>> linsolve([e1.expand(), e2.expand()], x, y, strict=False)
+    {(1, 1)}
```

Comment:
I'm still trying to think if there is a better way to define what should or shouldn't be allowed here.

Reply:
Maybe only allow the factor to contain generators that are not in syms. So `1/x` is generator `x` to the negative one power and would be disallowed by generator `x` while `sin(x)` would not be disallowed.

Reply:
```python
>>> def has_generator(self, s):
...     if self in s:
...         return True
...     elif self.is_Add or self.is_Mul:
...         return any(a.has_generator(s) for a in self.args)
...     elif self.is_Pow:
...         return self.base.has_generator(s) and not self.exp.has_xfree(s)
...     return False
...
>>> Basic.has_generator = has_generator
>>> (1/x).has_generator({x})
True
>>> (1/f(x)).has_generator({x})
False
>>> f(x).has_generator({x})
False
>>> (1+x).has_generator({x})
True
>>> x.has_generator({x**2})
False
>>> (x**2).has_generator({x})
True
>>> (2**x).has_generator({x**2})
False
>>> (x**y).has_generator({x**2})
False
```